### PR TITLE
ISPN-13512 Configuration converter

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-13512

ISPN-13512 Configuration converter
    
* Add CLI `config convert` command
* Complete server config serialization
* Proper serialization API for secret masking
* Implement JGroups stack serialization
* Delay JGroups stack initialization
* Fix parsing of YAML attribute lists
* Don't include implict authorization roles

### CODE REVIEW

**Review by ryanemerson — COMMENTED** [review]
The need for the server libs means that we either have todo the following for the native CLI and `infinispan/cli` image:

1. Don't support `config convert` as it won't be activated without the libs..
2. Include the required server libs in the native CLI.

I don't think 2 is desirable, so we can always direct users to utilise `infinispan/server` with `./bin/cli.sh` as the entrypoint if they want config conversion.

File: server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   private final MediaType type;
+
+   @Parameterized.Parameters(name = "{0} > [{2}]")
+   public static Iterable<Object[]> data() throws IOException {
+      Path configDir = Paths.get(System.getProperty("build.directory"), "test-classes", "configuration", "versions");
+      ServerConfigurationParserTest.createCredentialStore(configDir.getParent().resolve("credentials.pfx"), "secret");
+      return Files.list(configDir).flatMap(p -> types.stream().map(t -> new Object[]{p.getFileName(), p, t})).collect(Collectors.toList());
+   }
+
+   public ServerConfigurationSerializerTest(Path name, Path config, MediaType type) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java:47]: Ah, but it is :) 
`@Parameterized.Parameters(name = "{0} > [{2}]")`
That's the `{0}`

File: server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      }
+      try (ConfigurationWriter writer = ConfigurationWriter.to(baos).withType(type).clearTextSecrets(true).build()) {
+         registry.serialize(writer, holderBefore.getGlobalConfigurationBuilder().build(), configurations);
+      }
+      System.out.println(baos);
+      log.debug(baos);
+      ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+      ConfigurationBuilderHolder holderAfter = registry.parse(bais, null, type);
+      GlobalConfiguration globalConfigurationBefore = holderBefore.getGlobalConfigurationBuilder().build();
+      GlobalConfiguration globalConfigurationAfter = holderAfter.getGlobalConfigurationBuilder().build();
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java:71]: Yes, I need to fill that in.

File: cli/src/main/java/org/infinispan/cli/commands/CliCommand.java [hunk 3]
```diff
@@ -31,7 +31,7 @@ public CommandResult execute(ContextAwareCommandInvocation invocation) throws Co
       } catch (Throwable e) {
          // These are unhandled
          Throwable cause = Util.getRootCause(e);
-         invocation.getShell().writeln(ANSI.RED_TEXT + cause.getClass().getSimpleName() +": " + cause.getLocalizedMessage() + ANSI.DEFAULT_TEXT);
+         System.err.println(ANSI.RED_TEXT + cause.getClass().getSimpleName() +": " + cause.getLocalizedMessage() + ANSI.DEFAULT_TEXT);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/commands/CliCommand.java:34]: For some reason outputting through the shell is not working, effectively hiding error messages

File: commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java [hunk 4]
```diff
@@ -409,11 +409,27 @@ public ElementType nextElement() {
             }
          } else {
             // Read the attributes: they are indented relative to the element and have a value
-            while (next != null && next.indent > currentIndent && next.name != null && next.value != null) {
-               this.attributeNames.add(next.name);
-               this.attributeNamespaces.add(next.nsPrefix);
-               this.attributeValues.add(replaceProperties(next.value));
-               readNext();
+            while (next != null && next.indent > currentIndent && next.name != null) {// && next.value != null) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java]: Done

File: core/src/main/java/org/infinispan/configuration/global/JGroupsConfiguration.java [hunk 5]
```diff
@@ -54,6 +58,14 @@ public boolean isClustered() {
       return transport.get() != null;
    }
 
+   public JGroupsChannelConfigurator configurator(String name) {
+      return Stream.concat(stackConfigurations.stream(), stackFileConfigurations.stream())
+            .filter(s -> s.name().equals(name))
+            .map(s -> s.configurator())
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/global/JGroupsConfiguration.java]: Done

File: server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      ConfigurationBuilderHolder holderBefore = registry.parse(config);
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      Map<String, Configuration> configurations = new HashMap<>();
+      for (Map.Entry<String, ConfigurationBuilder> configuration : holderBefore.getNamedConfigurationBuilders().entrySet()) {
+         configurations.put(configuration.getKey(), configuration.getValue().build());
+      }
+      try (ConfigurationWriter writer = ConfigurationWriter.to(baos).withType(type).clearTextSecrets(true).build()) {
+         registry.serialize(writer, holderBefore.getGlobalConfigurationBuilder().build(), configurations);
+      }
+      System.out.println(baos);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/configuration/ServerConfigurationSerializerTest.java]: Done

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationSerializer.java [hunk 7]
```diff
@@ -302,6 +318,24 @@ private void writeConnector(ConfigurationWriter writer, RestServerConfiguration
...
 
    private void writeConnector(ConfigurationWriter writer, MemcachedServerConfiguration connector) {
+      if (connector.isImplicit()) {
+         return;
+      }
       connector.write(writer);
    }
+
+   public static AttributeSerializer<Supplier<char[]>> CREDENTIAL = (writer, name, value) -> {
+      String credential = writer.clearTextSecrets() ? new String(value.get()) : "***";
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationSerializer.java]: Done

**Comment by tristantarrant** [comment]
I have created [ISPN-13543 JSON Schema](https://issues.redhat.com/browse/ISPN-13543)

**Comment by tristantarrant** [comment]
> The need for the server libs means that we either have todo the following for the native CLI and `infinispan/cli` image:
> 
>     1. Don't support `config convert` as it won't be activated without the libs..
> 
>     2. Include the required server libs in the native CLI.
> 
> 
> I don't think 2 is desirable, so we can always direct users to utilise `infinispan/server` with `./bin/cli.sh` as the entrypoint if they want config conversion.

There is a third option: implement a full config conversion endpoint on the server (we currently have one which is limited to cache configs) and use it from the CLI when we don't have the parsers/serializers available [ISPN-13544 Server Config Converter Endpoint](https://issues.redhat.com/browse/ISPN-13544)

**Comment by ryanemerson** [comment]
> There is a third option: implement a full config conversion endpoint on the server (we currently have one which is limited to cache configs) and use it from the CLI when we don't have the parsers/serializers available [ISPN-13544 Server Config Converter Endpoint](https://issues.redhat.com/browse/ISPN-13544)

What use-case would that solve though? If the user wanted to use `infinispan/cli`, then in order to use ^ they would also have to  run `infinispan/server` at which point it will be easier for the user to just do:

```
cat infinispan.xml | docker run -i --entrypoint /bin/cli.sh quay.io/infinispan/server convert -f yaml 
```

**Comment by ryanemerson** [comment]
There are related failures.

**Comment by oraNod** [comment]
@tristantarrant I've pushed some doc suggestions. Also I couldn't get the short options to work like `-f` or `-o` and had to use `--format` and `--output`.

**Comment by tristantarrant** [comment]
I have fixed the short options as well as disabling the `ConfigurationUnitTest#testSchema` which is failing because of JGRP-2590

**Comment by tristantarrant** [comment]
@ryanemerson CI looks good now

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**